napi-inl.h

```diff
@@ -3627,6 +3686,357 @@ inline void AsyncWorker::OnWorkComplete(
+             initialThreadCount, context, finalizeCallback, data,
+             details::ThreadSafeFinalize<ContextType, Finalizer,
+                 FinalizerDataType>::WrapperWithDataAndContext);
+}
+
+
+inline ThreadSafeFunction::ThreadSafeFunction(
+    napi_env env, napi_threadsafe_function tsFunctionValue)
+  : _env(env),
+    _tsFunctionValue(tsFunctionValue) {
+}
+
+inline ThreadSafeFunction::~ThreadSafeFunction() {
+  Abort();
+}
+
+inline ThreadSafeFunction::ThreadSafeFunction(ThreadSafeFunction&& other)
+  : _env(other._env), _tsFunctionValue(other._tsFunctionValue) {
+  other._env = nullptr;
+  other._tsFunctionValue = nullptr;
+}
+
+inline ThreadSafeFunction& ThreadSafeFunction::operator =(
+    ThreadSafeFunction&& other) {
+  Abort();
+  _env = other._env;
+  _tsFunctionValue = other._tsFunctionValue;
+  other._env = nullptr;
+  other._tsFunctionValue = nullptr;
+  return *this;
+}
+
+inline ThreadSafeFunction::Status ThreadSafeFunction::BlockingCall() const {
+  return CallInternal(nullptr, napi_tsfn_blocking);
+}
+
+template <typename Callback>
+inline ThreadSafeFunction::Status ThreadSafeFunction::BlockingCall(
+    Callback callback) const {
+  return CallInternal(new CallbackWrapper(callback), napi_tsfn_blocking);
+}
+
+template <typename DataType, typename Callback>
+inline ThreadSafeFunction::Status ThreadSafeFunction::BlockingCall(
+    DataType* data, Callback callback) const {
+  auto wrapper = [data, callback](Env env, Function jsCallback) {
+    callback(env, jsCallback, data);
+  };
+  return CallInternal(new CallbackWrapper(wrapper), napi_tsfn_blocking);
+}
+
+inline ThreadSafeFunction::Status ThreadSafeFunction::NonBlockingCall() const {
+  return CallInternal(nullptr, napi_tsfn_nonblocking);
+}
+
+template <typename Callback>
+inline ThreadSafeFunction::Status ThreadSafeFunction::NonBlockingCall(
+    Callback callback) const {
+  return CallInternal(new CallbackWrapper(callback), napi_tsfn_nonblocking);
+}
+
+template <typename DataType, typename Callback>
+inline ThreadSafeFunction::Status ThreadSafeFunction::NonBlockingCall(
+    DataType* data, Callback callback) const {
+  auto wrapper = [data, callback](Env env, Function jsCallback) {
+    callback(env, jsCallback, data);
+  };
+  return CallInternal(new CallbackWrapper(wrapper), napi_tsfn_nonblocking);
+}
+
+inline bool ThreadSafeFunction::Acquire() const {
+  return napi_acquire_threadsafe_function(_tsFunctionValue) == napi_ok;
+}
+
+inline bool ThreadSafeFunction::Release() {
+  return napi_release_threadsafe_function(
+      _tsFunctionValue, napi_tsfn_release) == napi_ok;
+}
+
+inline bool ThreadSafeFunction::Abort() {
+  if (IsAborted()) {
+    return false;
+  }
+
+  napi_status status = napi_release_threadsafe_function(
+      _tsFunctionValue, napi_tsfn_abort);
+
+  _tsFunctionValue = nullptr;
+  _env = nullptr;
+
+  return status == napi_ok;
+}
+
+inline bool ThreadSafeFunction::IsAborted() const {
```

Comment:
I don't believe we can accommodate such an API, because the checks below are not thread-safe. The thread-safe way to check if a `napi_threadsafe_function` has been aborted is to attempt to release it and to receive `napi_invalid_arg`. I understand that if you abort the thread-safe function in one thread you need not release it in other threads, but the flag informing you that you need not release it cannot itself be stored in the object representing the thread-safe function, because that is not thread-safe.

The behaviour whereby a thread receives `napi_closing` when attempting to place a call to indicate that it need not release the function has to be implemented in the application.